File: 剑指offer/48.最长不含重复字符的子字符串.py
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        ans = 0

        window = defaultdict(int)

        while right < len(s):
            window[s[right]] += 1
            right += 1

            while left < right and window[s[right - 1]] == 2:
                window[s[left]] -= 1
                left += 1

            ans = max(ans, right - left)

        return ans

File: 剑指offer/48.最长不含重复字符的子字符串.cpp
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> window;
        int left = 0;
        int right = 0;
        int ans = 0;
        while (right < s.size()) {
            char add_c = s[right];
            window[add_c] += 1;
            right += 1;

            while (window[add_c] == 2) {
                window[s[left]] -= 1;
                left += 1;
            }

            ans = max(ans, right - left);
        }

        return ans;
    }
};